cpp/src/arrow/dataset/partition_test.cc

```diff
@@ -0,0 +1,137 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "arrow/dataset/api.h"
+#include "arrow/dataset/partition.h"
+#include "arrow/dataset/test_util.h"
+#include "arrow/filesystem/localfs.h"
+#include "arrow/filesystem/path_util.h"
+#include "arrow/status.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/util/io_util.h"
+#include "arrow/util/string_view.h"
+
+namespace arrow {
+namespace dataset {
+
+class TestPartitionScheme : public ::testing::Test {
+ public:
+  void AssertParseError(const std::string& path) {
+    std::string unconsumed;
+    std::shared_ptr<Expression> ignored;
+    ASSERT_RAISES(Invalid, scheme_->Parse(path, &unconsumed, &ignored));
+  }
+
+  void AssertParse(const std::string& expected_consumed,
+                   const std::string& expected_unconsumed,
+                   std::shared_ptr<Expression> expected) {
+    for (std::string suffix : {"", "/dat.parquet"}) {
+      std::string unconsumed;
+      std::shared_ptr<Expression> parsed;
+      ASSERT_OK(scheme_->Parse(expected_consumed + expected_unconsumed + suffix,
+                               &unconsumed, &parsed));
+
+      ASSERT_NE(parsed, nullptr);
+      ASSERT_EQ(expected_unconsumed + suffix, unconsumed);
+      ASSERT_TRUE(parsed->Equals(*expected)) << parsed->ToString() << "\n"
+                                             << expected->ToString();
+    }
+  }
+
+  void AssertParse(const std::string& expected_consumed,
+                   const std::string& expected_unconsumed, const Expression& expected) {
+    AssertParse(expected_consumed, expected_unconsumed, expected.Copy());
+  }
+
+ protected:
+  std::shared_ptr<PartitionScheme> scheme_;
+};
+
+TEST_F(TestPartitionScheme, Simple) {
+  auto expr = equal(field_ref("alpha"), scalar<int16_t>(3));
+  scheme_ = std::make_shared<SimplePartitionScheme>(expr);
+  AssertParse("", "/hello/world", expr);
+}
+
+TEST_F(TestPartitionScheme, Hive) {
+  scheme_ = std::make_shared<HivePartitionScheme>(
+      schema({field("alpha", int32()), field("beta", float32())}));
+
+  AssertParse("/alpha=0/beta=3.25", "", "alpha"_ == int32_t(0) and "beta"_ == 3.25f);
+  AssertParse("/beta=3.25/alpha=0", "", "beta"_ == 3.25f and "alpha"_ == int32_t(0));
+  AssertParse("/alpha=0", "", "alpha"_ == int32_t(0));
+  AssertParse("/beta=3.25", "", "beta"_ == 3.25f);
+
+  AssertParse("/alpha=0", "/unexpected/beta=3.25", "alpha"_ == int32_t(0));
```

Comment:
@nealrichardson Read as: given schema `alpha:int, beta:float`, the path `/alpha=0/unexpected/beta/3.25` will not be parsed beyond `/unexpected`, causing the beta condition to not be found.

Is this what you would expect?

Reply:
If this is Hive-style, I'll defer to whatever the Hive spec says, or whatever the consensus is among software that implements Hive-style partitioning.

With generic regular-expression parsing of partitions from file paths, you should be able to define an expression that allows "unexpected" path segments.

Reply:
See my comments in partition.h on `PartitionScheme::Parse`. IMHO, exposing a partial parser and unconsumed strings is too complex for few benefits. For complex usage, the user should be able to provide a custom partition scheme, likely backed by an implementation (R/Python) language lambda.